ga4gh/exceptions.py

```diff
@@ -0,0 +1,179 @@
+        it can be communicated back to the client.
+        """
+        error = protocol.GAException()
+        error.errorCode = self.getErrorCode()
+        error.message = self.getMessage()
+        return error
+
+#####################################################################
+#
+# Exceptions that occur in the normal operation of the server
+#
+#####################################################################
+
+
+class BadRequestException(ServerException):
+    """
+    A request that we don't like was sent to the server.
+    """
+    httpStatus = 400
+    message = "Bad request"
+
+
+class BadPageSizeException(BadRequestException):
+
+    message = "Request page size invalid"
+
+
+class BadPageTokenException(BadRequestException):
+
+    message = "Request page token invalid"
+
+
+class NotFoundException(ServerException):
+    """
+    The superclass of all exceptions in which some resource was not
+    found.
+    """
+    message = "A resource was not found"
+    httpStatus = 404
+
+
+class PathNotFoundException(NotFoundException):
+
+    message = "The request path was not found"
+
+
+class ObjectNotFoundException(NotFoundException):
+
+    message = "The requested object was not found"
+
+
+class UnsupportedMediaTypeException(ServerException):
+    httpStatus = 415
+    message = "Unsupported media type"
+
+
+class VersionNotSupportedException(NotFoundException):
+
+    message = "API version not supported"
+
+
+class MethodNotAllowedException(ServerException):
+
+    httpStatus = 405
+    message = "Method not allowed"
+
+
+class NotImplementedException(ServerException):
+    """
+    Exception raised when a part of the API has not been implemented.
+    """
+    httpStatus = 501
+
+    def __init__(self, message):
+        self.message = message
+
+
+class RequestValidationFailureException(ServerException):
+    """
+    A validation of the request data failed
+    """
+
+    def getMessage(self):
+        message = (
+            "Malformed request: JSON does not conform to the GA4GH"
+            "protocol version {}".format(protocol.version))
+        return message
+
+
+class CallSetNotInVariantSetException(NotFoundException):
+    """
+    Indicates a request was made for a callSet not in the actual variantSet
+    """
+
+    def __init__(self, callSetId, variantSetId):
+        self.callSetId = callSetId
+        self.variantSetId = variantSetId
+
+    def getMessage(self):
+        message = "callSet '{0}' not in variantSet '{1}'".format(
+            self.callSetId, self.variantSetId)
+        return message
+
+###############################################################
+#
+# Internal errors. These are exceptions that we regard as bugs.
+#
+###############################################################
+
+
+class ServerError(ServerException):
+    """
+    Superclass of all exceptions that indicate a bug has occured.
+    """
+    httpStatus = 500
+
+
+class ResponseValidationFailureException(ServerError):
```

Comment:
No, this is a different class of error. It's not a ServerError if someone sends us unusable crud, but it certainly is if we _return_ unusable crud.